plugins/inputs/redfish/redfish.go

```diff
@@ -0,0 +1,358 @@
+	UpperThresholdCritical int
+	UpperThresholdFatal    int
+	Status                 FansStatus
+}
+type FansStatus struct {
+	State  string
+	Health string
+}
+type Watt struct {
+	Name                 string
+	PowerInputWatts      float64
+	PowerCapacityWatts   float64
+	PowerOutputWatts     float64
+	LastPowerOutputWatts float64
+	Status               PowerStatus
+	LineInputVoltage     float64
+}
+type PowerStatus struct {
+	State  string
+	Health string
+}
+type Volt struct {
+	Name                   string
+	ReadingVolts           float64
+	UpperThresholdCritical float64
+	UpperThresholdFatal    float64
+	Status                 VoltStatus
+}
+type VoltStatus struct {
+	State  string
+	Health string
+}
+type Address struct {
+	PostalAddress PostalAddress
+	Placement     Placement
+}
+type PostalAddress struct {
+	DataCenter string
+	Room       string
+}
+type Placement struct {
+	Rack string
+	Row  string
+}
+type Redfish struct {
+	Host              string `toml:"host"`
+	BasicAuthUsername string `toml:"username"`
+	BasicAuthPassword string `toml:"password"`
+	Id                string `toml:"id"`
+	client            http.Client
+	tls.ClientConfig
+	Timeout internal.Duration `toml:"timeout"`
+}
+
+func (r *Redfish) Description() string {
+	return "Read CPU, Fans, Powersupply and Voltage metrics of Dell/HP hardware server through redfish APIs"
+}
+
+var redfishConfig = `
+  ##Server  OOB-IP
+  host = "192.0.0.1"
+
+  ##Username,  Password   for   hardware   server
+  username = "test"
+  password = "test"
+
+  ##Resource  Id   for   redfish   APIs
+  id="System.Embedded.1"
+
+  ##Optional TLS   Config, if not provided insecure skip verifies defaults to true
+  #tls_ca = "/etc/telegraf/ca.pem"
+  #tls_cert = "/etc/telegraf/cert.pem"
+  #tls_key = "/etc/telegraf/key.pem"
+  
+
+  ## Amount   of   time   allowed   to   complete   the   HTTP   request
+  # timeout = "5s"
+`
+
+func (r *Redfish) SampleConfig() string {
+	return redfishConfig
+}
+
+func (r *Redfish) Init() error {
+	tlsCfg, err := r.ClientConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+	if (len(r.ClientConfig.TLSCA) == 0) || (len(r.ClientConfig.TLSCert) == 0 && len(r.ClientConfig.TLSKey) == 0) {
+		var insecuretls tls.ClientConfig
+		insecuretls.InsecureSkipVerify = true
+		tlsCfg, err = insecuretls.TLSConfig()
+		if err != nil {
+			return err
+		}
+	}
+	r.client = http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsCfg,
+			Proxy:           http.ProxyFromEnvironment,
+		},
+		Timeout: r.Timeout.Duration,
+	}
+	return nil
+}
+
+func Severity(critical, fatal, val int) string {
+	severity := "NA"
+	if (critical != 0) || (fatal != 0) {
+		if (val >= fatal) && (fatal != 0) {
+			severity = "Fatal"
+		} else if (val >= critical) && (critical != 0) {
+			severity = "Critical"
+		} else {
+			severity = "OK"
+		}
+	}
+	return severity
+}
+
+func (r *Redfish) Gather(acc telegraf.Accumulator) error {
+	var url []string
+	var payload Payload
```

Comment:
Parsing all the data into a single Payload type makes it more difficult to determine where the various bits of data are sourced from.  It also increases the chances of value conflicts across resources and will make it difficult to support multiple Power or Thermal resources per ComputerSystem.  Instead of using this shared type we should parse each resource into its own dedicated type.

I suggest we relay out of the Gather flow as follows:

- Create a `func get(url, type struct) error` that will do requests and parse the json into the provided struct.  Having this function will help with the response body closing mentioned in the review.

- In Gather Get the ComputerSystem, Chassis resources.  In the future it would be nice to follow the @odata.i links in case a system has multiple Thermal or Power resources, or the ID is not shared for all resources, but it is not required now.

- Next get the Thermal resource, then I suggest another function that creates the tags and fields using it and the Chassis and ComputerSystem types, and call AddFields.

- Finally get the Power resource, also in another function create the tags and fields using it and the Chassis and ComputerSystem types, and call AddFields.

Reply:
About parsing data into a single Payload was suggested by Steven the other reviewer, previously the json data was parsed into different structs, What shall we do now?

Reply:
Okay, don't worry about doing this refactor.